engine/access/ingestion/version/version_control.go

```diff
@@ -0,0 +1,281 @@
+
+// NoHeight represents the maximum possible height for blocks.
+var NoHeight = uint64(math.MaxUint64)
+
+// VersionControl manages the version control system for the node.
+// It consumes BlockFinalized events and updates the node's version control based on the latest version beacon.
+//
+// VersionControl implements the protocol.Consumer and component.Component interfaces.
+type VersionControl struct {
+	// Noop implements the protocol.Consumer interface with no operations.
+	psEvents.Noop
+	sync.RWMutex
+	component.Component
+
+	log zerolog.Logger
+	// Storage
+	versionBeacons storage.VersionBeacons
+
+	// nodeVersion stores the node's current version.
+	// It could be nil if the node version is not available.
+	nodeVersion *semver.Version
+	// versionBeacon stores the last handled version beacon.
+	versionBeacon *flow.SealedVersionBeacon
+
+	// consumers stores the list of consumers for version updates.
+	consumers []VersionControlConsumer
+
+	// blockFinalizedChan is a channel for receiving block finalized events.
+	blockFinalizedChan chan *flow.Header
+
+	// startHeight and endHeight define the height boundaries for version compatibility.
+	startHeight uint64
+	endHeight   uint64
+}
+
+var _ protocol.Consumer = (*VersionControl)(nil)
+var _ component.Component = (*VersionControl)(nil)
+
+// NewVersionControl creates a new VersionControl instance.
+//
+// We currently have no strong guarantee that the node version is a valid semver.
+// See build.SemverV2 for more details. That is why nil is a valid input for node version.
+// Without a node version, the stop control can still be used for manual stopping.
+func NewVersionControl(
+	log zerolog.Logger,
+	versionBeacons storage.VersionBeacons,
+	nodeVersion *semver.Version,
+	latestFinalizedBlock *flow.Header,
+) *VersionControl {
+	// blockFinalizedChan is buffered to handle block finalized events.
+	blockFinalizedChan := make(chan *flow.Header, 1000)
+
+	vc := &VersionControl{
+		log: log.With().
+			Str("component", "version_control").
+			Logger(),
+
+		blockFinalizedChan: blockFinalizedChan,
+		nodeVersion:        nodeVersion,
+		versionBeacons:     versionBeacons,
+		startHeight:        NoHeight,
+		endHeight:          NoHeight,
+	}
+
+	if vc.nodeVersion != nil {
+		log = log.With().
+			Stringer("node_version", vc.nodeVersion).
+			Logger()
+	}
+
+	log.Info().Msg("Created")
+
+	// Setup component manager for handling worker functions.
+	cm := component.NewComponentManagerBuilder()
+	cm.AddWorker(vc.processEvents)
+	cm.AddWorker(func(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+		vc.checkInitialVersionBeacon(ctx, ready, latestFinalizedBlock)
+	})
+
+	vc.Component = cm.Build()
+
+	return vc
+}
+
+// checkInitialVersionBeacon checks the initial version beacon at the latest finalized block.
+// It ensures the component is not ready until the initial version beacon is checked.
+func (v *VersionControl) checkInitialVersionBeacon(
```

Comment:
this doesn't quite capture the logic needed. A node that's starting up can be in a few conditions:

* Compatible with all blocks in the network -> there has not been any version beacons during the spork
* Compatible with all blocks before some (finalized or future) height
* Compatible with all blocks after some past height
* Compatible with all blocks between 2 heights

It's also possible that the node has been offline, and doesn't yet know about some version beacon events.

this needs to scan through the version beacon events the node has in it's db, until we know the start and end heights, or that there is no start/end height.